File: Sources/VocalFlow/LLMService.swift
import Foundation
import os.log

private let llmLogger = Logger(subsystem: "com.vocalflow.app", category: "llm")

enum LLMProvider: String, CaseIterable, Identifiable {
    case groq        = "groq"
    case openRouter  = "open_router"

    var id: String { rawValue }

    var displayName: String {
        switch self {
        case .groq:       return "Groq"
        case .openRouter: return "OpenRouter"
        }
    }

    var baseURL: URL {
        switch self {
        case .groq:       return URL(staticString: "https://api.groq.com/openai/v1")
        case .openRouter: return URL(staticString: "https://openrouter.ai/api/v1")
        }
    }

    var modelsURL: URL          { baseURL.appendingPathComponent("models") }
    var chatCompletionsURL: URL { baseURL.appendingPathComponent("chat/completions") }

    var signupURL: URL {
        switch self {
        case .groq:       return URL(staticString: "https://console.groq.com/keys")
        case .openRouter: return URL(staticString: "https://openrouter.ai/keys")
        }
    }

    var keychainKey: String {
        switch self {
        case .groq:       return "groq_api_key"
        case .openRouter: return "openrouter_api_key"
        }
    }
}

struct LLMModel: Identifiable, Hashable {
    let id: String
    let displayName: String
}

struct LLMProcessingOptions {
    var codeMix: String?        // nil = disabled; value = mixType e.g. "Hinglish"
    var fixSpelling: Bool
    var fixGrammar: Bool
    var targetLanguage: String? // nil = disabled; value = e.g. "French"
    var customPrompt: String?   // nil/empty = disabled; user-supplied bias prepended to system prompt

    var hasAnyStep: Bool {
        if codeMix != nil || fixSpelling || fixGrammar || targetLanguage != nil { return true }
        let trimmed = customPrompt?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        return !trimmed.isEmpty
    }
}

class LLMService {
    /// Per-request timeout for chat completions. Default URLSession is 60s, but some
    /// OpenRouter models (Anthropic Sonnet, large Gemini) routinely take 30+ seconds —
    /// 90s leaves headroom without hanging the UI forever.
    var requestTimeout: TimeInterval = 90

    /// Delay before the single retry on 429/5xx / network blips.
    var retryDelay: TimeInterval = 0.25

    func fetchModels(provider: LLMProvider, apiKey: String) async throws -> [LLMModel] {
        guard !apiKey.isEmpty else { throw APIError.missingKey }

        var request = URLRequest(url: provider.modelsURL)
        request.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")

        let (data, response) = try await performRequest(request)
        let httpStatus = (response as? HTTPURLResponse)?.statusCode ?? 0
        guard (200..<300).contains(httpStatus) else {
            let body = String(data: data, encoding: .utf8)
            llmLogger.error("\(provider.displayName) /models HTTP \(httpStatus): \(body ?? "<binary>")")
            throw APIError.http(status: httpStatus, body: body)
        }
        guard let root = try? JSONDecoder().decode(LLMModelsResponse.self, from: data) else {
            let body = String(data: data, encoding: .utf8)
            llmLogger.error("\(provider.displayName) /models decode failed: \(body ?? "<binary>")")
            throw APIError.decoding
        }
        let models = (root.data ?? [])
            .map { LLMModel(id: $0.id, displayName: $0.name ?? $0.id) }
            .sorted { $0.displayName.lowercased() < $1.displayName.lowercased() }
        llmLogger.info("\(provider.displayName) /models returned \(models.count) entries")
        return models
    }

    func processText(
        _ text: String,
        options: LLMProcessingOptions,
        provider: LLMProvider,
        apiKey: String,
        model: String
    ) async throws -> String {
        guard let systemPrompt = Self.buildSystemPrompt(for: options) else { return text }
        guard !apiKey.isEmpty else { throw APIError.missingKey }
        guard !model.isEmpty else { throw APIError.missingModel }

        var request = URLRequest(url: provider.chatCompletionsURL)
        request.httpMethod = "POST"
        request.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.timeoutInterval = requestTimeout
        if provider == .openRouter {
            request.setValue("https://github.com/Vocallabsai/vocalflow", forHTTPHeaderField: "HTTP-Referer")
            request.setValue("VocalFlow", forHTTPHeaderField: "X-Title")
        }

        let body: [String: Any] = [
            "model": model,
            "messages": [
                ["role": "system", "content": systemPrompt],
                ["role": "user",   "content": text]
            ],
            "temperature": 0
        ]

        guard let httpBody = try? JSONSerialization.data(withJSONObject: body) else {
            llmLogger.error("Failed to encode chat completion request body")
            throw APIError.encoding
        }
        request.httpBody = httpBody

        return try await sendChatCompletion(request: request, provider: provider, allowRetry: true)
    }

    /// Pure prompt assembly. Extracted as `static` so it's deterministic given options
    /// and unit-testable without spinning up a real network. Returns `nil` when no
    /// processing steps are enabled (caller short-circuits to passthrough).
    static func buildSystemPrompt(for options: LLMProcessingOptions) -> String? {
        guard options.hasAnyStep else { return nil }

        var instructions: [String] = []
        var stepNumber = 1

        if let mixType = options.codeMix {
            instructions.append("\(stepNumber). The input is in \(mixType). Transliterate any non-Roman script (such as Devanagari, Tamil, etc.) to Roman script. Keep English words as-is. Do not translate — preserve the original meaning in mixed form.")
            stepNumber += 1
        }
        if options.fixSpelling {
            instructions.append("\(stepNumber). Fix any spelling mistakes. Do not change meaning or structure.")
            stepNumber += 1
        }
        if options.fixGrammar {
            instructions.append("\(stepNumber). Fix any grammar mistakes. Do not change meaning or add content.")
            stepNumber += 1
        }
        if let lang = options.targetLanguage {
            instructions.append("\(stepNumber). Translate the entire text to \(lang). Every word must be in \(lang).")
        }

        let stepsBlock: String? = instructions.isEmpty ? nil :
            "Process the following text by applying these steps in order:\n"
            + instructions.joined(separator: "\n")

        let trimmedCustom = options.customPrompt?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        let customBlock: String? = trimmedCustom.isEmpty ? nil : trimmedCustom

        let blocks = [customBlock, stepsBlock, "Return only the final processed text with no explanation."]
            .compactMap { $0 }
        return blocks.joined(separator: "\n\n")
    }

    // MARK: - Internal helpers

    private func sendChatCompletion(
        request: URLRequest,
        provider: LLMProvider,
        allowRetry: Bool
    ) async throws -> String {
        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await performRequest(request)
        } catch let error as APIError {
            if allowRetry, case .network = error {
                llmLogger.info("Retrying \(provider.displayName) /chat/completions after network error")
                try? await Task.sleep(nanoseconds: UInt64(retryDelay * 1_000_000_000))
                return try await sendChatCompletion(request: request, provider: provider, allowRetry: false)
            }
            throw error
        }

        let httpStatus = (response as? HTTPURLResponse)?.statusCode ?? 0
        if !(200..<300).contains(httpStatus) {
            let bodyStr = String(data: data, encoding: .utf8)
            llmLogger.error("\(provider.displayName) /chat/completions HTTP \(httpStatus): \(bodyStr ?? "<binary>")")
            if allowRetry && (httpStatus == 429 || (500..<600).contains(httpStatus)) {
                llmLogger.info("Retrying \(provider.displayName) /chat/completions after HTTP \(httpStatus)")
                try? await Task.sleep(nanoseconds: UInt64(retryDelay * 1_000_000_000))
                return try await sendChatCompletion(request: request, provider: provider, allowRetry: false)
            }
            throw APIError.http(status: httpStatus, body: bodyStr)
        }

        guard let root = try? JSONDecoder().decode(LLMChatResponse.self, from: data),
              let result = root.choices?.first?.message?.content,
              !result.isEmpty else {
            let bodyStr = String(data: data, encoding: .utf8)
            llmLogger.error("\(provider.displayName) /chat/completions decode failed: \(bodyStr ?? "<binary>")")
            throw APIError.decoding
        }
        return Self.stripReasoning(result)
    }

    /// Strip reasoning/thinking blocks that some models (DeepSeek R1, Qwen QwQ, etc.) emit
    /// inline in the assistant content. Handles `<think>...</think>`, `<thinking>...</thinking>`,
    /// and `<reasoning>...</reasoning>` tags. Case-insensitive, multiline. If an opening tag
    /// has no closing tag (streamed cut-off), drop everything up to the last seen opener.
    static func stripReasoning(_ text: String) -> String {
        let tags = ["think", "thinking", "reasoning"]
        var output = text
        for tag in tags {
            let pattern = "<\\s*\(tag)\\s*>[\\s\\S]*?<\\s*/\\s*\(tag)\\s*>"
            if let regex = try? NSRegularExpression(pattern: pattern, options: [.caseInsensitive]) {
                let range = NSRange(output.startIndex..., in: output)
                output = regex.stringByReplacingMatches(in: output, options: [], range: range, withTemplate: "")
            }
            // Unbalanced opener: keep only text after the last opening tag.
            let openPattern = "<\\s*\(tag)\\s*>"
            if let openRegex = try? NSRegularExpression(pattern: openPattern, options: [.caseInsensitive]) {
                let range = NSRange(output.startIndex..., in: output)
                if let last = openRegex.matches(in: output, options: [], range: range).last,
                   let swiftRange = Range(last.range, in: output) {
                    output = String(output[swiftRange.upperBound...])
                }
            }
        }
        return output.trimmingCharacters(in: .whitespacesAndNewlines)
    }

    private func performRequest(_ request: URLRequest) async throws -> (Data, URLResponse) {
        do {
            return try await URLSession.shared.data(for: request)
        } catch {
            throw APIError.network(error.localizedDescription)
        }
    }
}

// MARK: - Response Models

private struct LLMModelsResponse: Codable {
    let data: [LLMModelEntry]?
}

private struct LLMModelEntry: Codable {
    let id: String
    let name: String?
}

private struct LLMChatResponse: Codable {
    let choices: [LLMChoice]?
}

private struct LLMChoice: Codable {
    let message: LLMMessage?
}

private struct LLMMessage: Codable {
    let content: String?
}

File: windows/VocalFlow/Core/LlmService.cs
using System.Diagnostics;
using System.Net.Http;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Text.RegularExpressions;

namespace VocalFlow.Core;

/// <summary>
/// Calls Groq/OpenRouter OpenAI-compatible chat completions to post-process a transcript,
/// and lists available models. Port of the macOS LlmService.
/// </summary>
public sealed class LlmService
{
    private static readonly HttpClient ModelsHttp = new();

    /// <summary>Per-request timeout for chat completions (some OpenRouter models take 30s+).</summary>
    public TimeSpan RequestTimeout { get; set; } = TimeSpan.FromSeconds(90);

    /// <summary>Delay before the single retry on 429/5xx / network blips.</summary>
    public TimeSpan RetryDelay { get; set; } = TimeSpan.FromMilliseconds(250);

    public async Task<IReadOnlyList<LlmModel>> FetchModelsAsync(LlmProvider provider, string apiKey)
    {
        if (string.IsNullOrEmpty(apiKey)) throw ApiException.MissingKey();

        using var request = new HttpRequestMessage(HttpMethod.Get, provider.ModelsUrl());
        request.Headers.TryAddWithoutValidation("Authorization", $"Bearer {apiKey}");

        var (data, status) = await PerformAsync(ModelsHttp, request).ConfigureAwait(false);
        if (status < 200 || status >= 300)
            throw ApiException.Http(status, data);

        LlmModelsResponse? root;
        try { root = JsonSerializer.Deserialize<LlmModelsResponse>(data); }
        catch { throw ApiException.Decoding(); }
        if (root == null) throw ApiException.Decoding();

        return (root.Data ?? new())
            .Select(m => new LlmModel(m.Id, m.Name ?? m.Id))
            .OrderBy(m => m.DisplayName.ToLowerInvariant(), StringComparer.Ordinal)
            .ToList();
    }

    public async Task<string> ProcessTextAsync(
        string text, LlmProcessingOptions options, LlmProvider provider, string apiKey, string model)
    {
        var systemPrompt = BuildSystemPrompt(options);
        if (systemPrompt == null) return text;
        if (string.IsNullOrEmpty(apiKey)) throw ApiException.MissingKey();
        if (string.IsNullOrEmpty(model)) throw ApiException.MissingModel();

        var body = new
        {
            model,
            messages = new[]
            {
                new { role = "system", content = systemPrompt },
                new { role = "user", content = text },
            },
            temperature = 0,
        };

        string json;
        try { json = JsonSerializer.Serialize(body); }
        catch { throw ApiException.Encoding(); }

        // A fresh client per call so we can apply the long per-request timeout.
        using var client = new HttpClient { Timeout = RequestTimeout };
        return await SendChatCompletionAsync(client, provider, apiKey, json, allowRetry: true).ConfigureAwait(false);
    }

    /// <summary>
    /// Pure prompt assembly. Deterministic given options; returns null when no steps are enabled.
    /// </summary>
    public static string? BuildSystemPrompt(LlmProcessingOptions options)
    {
        if (!options.HasAnyStep) return null;

        var instructions = new List<string>();
        int step = 1;

        if (options.CodeMix is { } mixType)
        {
            instructions.Add($"{step}. The input is in {mixType}. Transliterate any non-Roman script (such as Devanagari, Tamil, etc.) to Roman script. Keep English words as-is. Do not translate — preserve the original meaning in mixed form.");
            step++;
        }
        if (options.FixSpelling)
        {
            instructions.Add($"{step}. Fix any spelling mistakes. Do not change meaning or structure.");
            step++;
        }
        if (options.FixGrammar)
        {
            instructions.Add($"{step}. Fix any grammar mistakes. Do not change meaning or add content.");
            step++;
        }
        if (options.TargetLanguage is { } lang)
        {
            instructions.Add($"{step}. Translate the entire text to {lang}. Every word must be in {lang}.");
        }

        string? stepsBlock = instructions.Count == 0
            ? null
            : "Process the following text by applying these steps in order:\n" + string.Join("\n", instructions);

        var trimmedCustom = options.CustomPrompt?.Trim() ?? "";
        string? customBlock = trimmedCustom.Length == 0 ? null : trimmedCustom;

        var blocks = new[] { customBlock, stepsBlock, "Return only the final processed text with no explanation." }
            .Where(b => b != null);
        return string.Join("\n\n", blocks);
    }

    private async Task<string> SendChatCompletionAsync(
        HttpClient client, LlmProvider provider, string apiKey, string json, bool allowRetry)
    {
        using var request = new HttpRequestMessage(HttpMethod.Post, provider.ChatCompletionsUrl());
        request.Headers.TryAddWithoutValidation("Authorization", $"Bearer {apiKey}");
        request.Content = new StringContent(json, Encoding.UTF8, "application/json");
        if (provider == LlmProvider.OpenRouter)
        {
            request.Headers.TryAddWithoutValidation("HTTP-Referer", "https://github.com/Vocallabsai/vocalflow");
            request.Headers.TryAddWithoutValidation("X-Title", "VocalFlow");
        }

        int status;
        string data;
        try
        {
            (data, status) = await PerformAsync(client, request).ConfigureAwait(false);
        }
        catch (ApiException e) when (allowRetry && e.ErrorKind == ApiException.Kind.Network)
        {
            await Task.Delay(RetryDelay).ConfigureAwait(false);
            return await SendChatCompletionAsync(client, provider, apiKey, json, allowRetry: false).ConfigureAwait(false);
        }

        if (status < 200 || status >= 300)
        {
            Debug.WriteLine($"[llm] {provider.DisplayName()} /chat/completions HTTP {status}: {data}");
            if (allowRetry && (status == 429 || (status >= 500 && status < 600)))
            {
                await Task.Delay(RetryDelay).ConfigureAwait(false);
                return await SendChatCompletionAsync(client, provider, apiKey, json, allowRetry: false).ConfigureAwait(false);
            }
            throw ApiException.Http(status, data);
        }

        LlmChatResponse? root;
        try { root = JsonSerializer.Deserialize<LlmChatResponse>(data); }
        catch { throw ApiException.Decoding(); }
        var result = root?.Choices is { Count: > 0 } ch ? ch[0].Message?.Content : null;
        if (string.IsNullOrEmpty(result))
            throw ApiException.Decoding();

        return StripReasoning(result!);
    }

    /// <summary>
    /// Strip reasoning/thinking blocks some models emit inline. Handles &lt;think&gt;, &lt;thinking&gt;,
    /// and &lt;reasoning&gt; tags (case-insensitive). For an unbalanced opener, keep only text after
    /// the last opening tag.
    /// </summary>
    public static string StripReasoning(string text)
    {
        string[] tags = { "think", "thinking", "reasoning" };
        var output = text;
        foreach (var tag in tags)
        {
            var pattern = $"<\\s*{tag}\\s*>[\\s\\S]*?<\\s*/\\s*{tag}\\s*>";
            output = Regex.Replace(output, pattern, "", RegexOptions.IgnoreCase);

            var openPattern = $"<\\s*{tag}\\s*>";
            var matches = Regex.Matches(output, openPattern, RegexOptions.IgnoreCase);
            if (matches.Count > 0)
            {
                var last = matches[^1];
                output = output[(last.Index + last.Length)..];
            }
        }
        return output.Trim();
    }

    private static async Task<(string body, int status)> PerformAsync(HttpClient client, HttpRequestMessage request)
    {
        try
        {
            var response = await client.SendAsync(request).ConfigureAwait(false);
            var body = await response.Content.ReadAsStringAsync().ConfigureAwait(false);
            return (body, (int)response.StatusCode);
        }
        catch (Exception e)
        {
            throw ApiException.Network(e.Message);
        }
    }

    // MARK: - Response DTOs

    private sealed class LlmModelsResponse
    {
        [JsonPropertyName("data")] public List<LlmModelEntry>? Data { get; set; }
    }

    private sealed class LlmModelEntry
    {
        [JsonPropertyName("id")] public string Id { get; set; } = "";
        [JsonPropertyName("name")] public string? Name { get; set; }
    }

    private sealed class LlmChatResponse
    {
        [JsonPropertyName("choices")] public List<LlmChoice>? Choices { get; set; }
    }

    private sealed class LlmChoice
    {
        [JsonPropertyName("message")] public LlmMessage? Message { get; set; }
    }

    private sealed class LlmMessage
    {
        [JsonPropertyName("content")] public string? Content { get; set; }
    }
}
